core_v14/src/decoder/decode_type.rs

```diff
@@ -0,0 +1,484 @@
+	decode_fields(data, ty.fields(), types)
+}
+
+fn decode_variant_type(
+	data: &mut &[u8],
+	ty: &TypeDefVariant<PortableForm>,
+	types: &PortableRegistry,
+) -> Result<Variant, DecodeTypeError> {
+	let index = *data.get(0).ok_or(DecodeTypeError::Eof)?;
+	*data = &data[1..];
+
+	// Does a variant exist with the index we're looking for?
+	let variant = ty
+		.variants()
+		.iter()
+		.find(|v| v.index() == index)
+		.ok_or_else(|| DecodeTypeError::VariantNotFound(index, ty.clone()))?;
+
+	let fields = decode_fields(data, variant.fields(), types)?;
+	Ok(Variant { name: variant.name().clone(), values: fields })
+}
+
+/// Variant and Composite types both have fields; this will decode them into values.
+fn decode_fields(
+	data: &mut &[u8],
+	fields: &[Field<PortableForm>],
+	types: &PortableRegistry,
+) -> Result<Composite, DecodeTypeError> {
+	let are_named = fields.iter().any(|f| f.name().is_some());
+	let named_field_vals = fields.iter().map(|f| {
+		let name = f.name().cloned().unwrap_or_default();
+		decode_type_by_id(data, f.ty(), types).map(|val| (name, val))
+	});
+
+	if are_named {
+		let vals = named_field_vals.collect::<Result<_, _>>()?;
+		Ok(Composite::Named(vals))
+	} else {
+		let vals = named_field_vals.map(|r| r.map(|(_, v)| v)).collect::<Result<_, _>>()?;
+		Ok(Composite::Unnamed(vals))
+	}
+}
+
+fn decode_sequence_type(
+	data: &mut &[u8],
+	ty: &TypeDefSequence<PortableForm>,
+	types: &PortableRegistry,
+) -> Result<Composite, DecodeTypeError> {
+	// We assume that the sequence is preceeded by a compact encoded length, so that
+	// we know how many values to try pulling out of the data.
+	let len = Compact::<u64>::decode(data)?;
+	let values: Vec<_> =
+		(0..len.0).map(|_| decode_type_by_id(data, ty.type_param(), types)).collect::<Result<_, _>>()?;
+
+	Ok(Composite::Unnamed(values))
+}
+
+fn decode_array_type(
+	data: &mut &[u8],
+	ty: &TypeDefArray<PortableForm>,
+	types: &PortableRegistry,
+) -> Result<Composite, DecodeTypeError> {
+	// The length is known based on the type we want to decode into, so we pull out the number of items according
+	// to that, and don't need a length to exist in the SCALE encoded bytes
+	let values: Vec<_> =
+		(0..ty.len()).map(|_| decode_type_by_id(data, ty.type_param(), types)).collect::<Result<_, _>>()?;
+
+	Ok(Composite::Unnamed(values))
+}
+
+fn decode_tuple_type(
+	data: &mut &[u8],
+	ty: &TypeDefTuple<PortableForm>,
+	types: &PortableRegistry,
+) -> Result<Composite, DecodeTypeError> {
+	let values: Vec<_> = ty.fields().iter().map(|f| decode_type_by_id(data, f, types)).collect::<Result<_, _>>()?;
+
+	Ok(Composite::Unnamed(values))
+}
+
+fn decode_primitive_type(data: &mut &[u8], ty: &TypeDefPrimitive) -> Result<Primitive, DecodeTypeError> {
+	let val = match ty {
+		TypeDefPrimitive::Bool => Primitive::Bool(bool::decode(data)?),
+		TypeDefPrimitive::Char => {
+			// [jsdw] TODO: There isn't a `char::decode`. Why? Is it wrong to use u32 or is there a more "proper" way?
```

Comment:
https://docs.substrate.io/v3/advanced/scale-codec/

Strings are just encoded as an array of u8 so I it is possible there is no notion of `char`. I'm guessing this is something that scale-info added but which is not present in scale-codec

Reply:
It seems that way; I want to handle the scale info type properly, but since scale-codec isn.t implemented for `char` I'm not actually sure what the expectation is (so I'm just assuming that char will be a 32bit value like in Rust!)